### PR TITLE
Add Textline with special encoding

### PR DESCRIPTION
**Pull request by kcyang**
I like to add the function that adding text line with special Codepage encoding.

### CODE REVIEW

**Comment by JesperSchulz** [comment]
Thank you for your contribution! I will look into your code asap and get back to you if I have any questions or suggestions!

File: BaseApp/COD57000.TXT [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+  }
+  CODE
+  {
+    VAR
+      Encoding@1000 : DotNet "'mscorlib, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'.System.Text.Encoding";
+      StreamWriter@1001 : DotNet "'mscorlib, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'.System.IO.StreamWriter";
+
+    [External]
+    PROCEDURE AddTextLineWithEncoding@2(LineText@1000 : Text;VAR OutStream_@1001 : OutStream;CodePageNo@1002 : Integer);
```
Comment (grobyns) [inline BaseApp/COD57000.TXT]: Hi,

Please put the Encoding dotnet variable in its own wrapper and add that encoding wrapper codeunit as the parameter here. The reason is that if someone else wants to initialize the StreamWriter with other types of encoding ) i.e. not using a CodePageNo then they can.


The code would look something like:


[External]
PROCEDURE StreamWriter@1(OutStream@1000 : OutStream;DotNet_Encoding@1001 : Codeunit 30xx);
VAR
  Encoding@1002 : DotNet "'mscorlib'.System.Text.Encoding";
BEGIN
  DotNet_Encoding.GetEncoding(Encoding);
  StreamWriter := StreamWriter.StreamWriter(OutStream,Encoding);
END;

File: BaseApp/COD57000.TXT [hunk 2]
```diff
@@ -0,0 +1,40 @@
+OBJECT Codeunit 57000 DotNet Add Text Encoding
+{
+  OBJECT-PROPERTIES
+  {
+    Date=18-06-14;
```
Comment (JesperSchulz) [inline BaseApp/COD57000.TXT]: Please clear date, time and remove modified flag.

File: BaseApp/COD57000.TXT [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  PROPERTIES
+  {
+    OnRun=BEGIN
+          END;
+
+  }
+  CODE
+  {
+    VAR
+      Encoding@1000 : DotNet "'mscorlib, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'.System.Text.Encoding";
```
Comment (JesperSchulz) [inline BaseApp/COD57000.TXT]: As per @grobyns comment, these wrapper should only every have 1 global variable. You can see more examples in COD3000..COD3024 in this repo.

File: BaseApp/COD57000.TXT [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      StreamWriter@1001 : DotNet "'mscorlib, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'.System.IO.StreamWriter";
+
+    [External]
+    PROCEDURE AddTextLineWithEncoding@2(LineText@1000 : Text;VAR OutStream_@1001 : OutStream;CodePageNo@1002 : Integer);
+    BEGIN
+
+      StreamWriter := StreamWriter.StreamWriter(OutStream_,Encoding.GetEncoding(CodePageNo));
+      StreamWriter.WriteLine(LineText);
```
Comment (JesperSchulz) [inline BaseApp/COD57000.TXT]: All of these methods should be wrapped in individual functions. You would have to create a constructor, a WriteLine function, a Flush functions, a Close Function and a Dispose function.

File: BaseApp/COD57000.TXT [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      StreamWriter.Flush;
+      StreamWriter.Close;
+      StreamWriter.Dispose;
+    END;
+
+    BEGIN
+    {
+      System.IO.StreamWriter.'mscorlib, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
+      System.Text.Encoding.'mscorlib, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'
```
Comment (JesperSchulz) [inline BaseApp/COD57000.TXT]: Please remove these comments.

File: BaseApp/COD57000.TXT [hunk 6]
```diff
@@ -0,0 +1,40 @@
+OBJECT Codeunit 57000 DotNet Add Text Encoding
```
Comment (JesperSchulz) [inline BaseApp/COD57000.TXT]: Name should be DotNet_StreamWriter.

File: BaseApp/COD57000.TXT (hunk 1 above)
Comment (JesperSchulz) [inline BaseApp/COD57000.TXT]: Why OutStream_? What is the underscore for? :-)

File: BaseApp/COD57000.TXT (hunk 2 above)
Comment (kcyang) [reply BaseApp/COD57000.TXT]: cleared and modified.

File: BaseApp/COD57000.TXT (hunk 3 above)
Comment (kcyang) [reply BaseApp/COD57000.TXT]: Ok, It's applied.

File: BaseApp/COD57000.TXT (hunk 4 above)
Comment (kcyang) [reply BaseApp/COD57000.TXT]: It's been applied.

File: BaseApp/COD57000.TXT (hunk 5 above)
Comment (kcyang) [reply BaseApp/COD57000.TXT]: Removed.

File: BaseApp/COD57000.TXT (hunk 6 above)
Comment (kcyang) [reply BaseApp/COD57000.TXT]: Changed.

File: BaseApp/COD57000.TXT (hunk 1 above)
Comment (kcyang) [reply BaseApp/COD57000.TXT]: I've change the name. ;)

**Comment by kcyang** [comment]
I'd modified these as you requested changes. Could you please review again?

File: BaseApp/COD57000.TXT [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+      DotNet_StreamWriter@1001 : DotNet "'mscorlib, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'.System.IO.StreamWriter";
+
+    [External]
+    PROCEDURE WriteLine@2(LineText@1000 : Text);
+    BEGIN
+      DotNet_StreamWriter.WriteLine(LineText);
+    END;
+
+    [External]
+    PROCEDURE StreamWriter@1(VAR OutStream@1000 : OutStream;CodePageNo@1001 : Integer);
```
[the remark this answers is not in the document]
Comment (JesperSchulz) [reply BaseApp/COD57000.TXT]: Other than this, your changes look good! As soon as this is fixed, I believe your code is solid and ready to be submitted.

**Comment by kcyang** [comment]
Fixed! and add the DotNet_Encoding , I hope this is the last ;)

**Comment by JesperSchulz** [comment]
Looks great! You are officially the first partner who has added a .NET wrapper in this format :-) Very well done! I will approve the pull request and submit your changes to our master branch to be released with the fall release. Let me know if you need your changes to be backported to the current release. More info about our branching strategy can be found here:
https://blogs.msdn.microsoft.com/nav/2018/05/03/al-developer-previews-multiple-releases-and-github/

**Comment by JesperSchulz** [comment]
I will have to re-ID your codeunits, but I guess that is fine :-)

**Comment by JesperSchulz** [comment]
Re-IDed through bb5931a1dc7265a30c98f9517bbddf40e589ef4c.

**Comment by kcyang** [comment]
@JesperSchulz Hi, Schulz. if possible, could you please apply to the current release or preview? I have prepare the localization package. if so, let me know applied release version.

**Comment by JesperSchulz** [comment]
I will backport the fix asap. Expected GA will be end of July.

**Pull request merged**